### DISCUSSION TITLE
How to use node loader with npx?

### DISCUSSION THREAD

**Discussion by Maxim-Mazurok** [npm/feedback#564] [CLI] [answered]
This is a more general question to a more specific question from here: https://github.com/TypeStrong/ts-node/discussions/1463

I want to use `node --loader ts-node/esm` but I don't want to install `ts-node`.
I want to do something like this: `npx --package ts-node --call "node --loader ts-node/esm"`, but this doesn't work and says "Cannot find package 'ts-node'":
```text
C:\Users\maxim>npx --package ts-node --call "node --loader ts-node/esm"
(node:11792) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
node:internal/errors:464
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'ts-node' imported from C:\Users\maxim\
    at new NodeError (node:internal/errors:371:5)
    at packageResolve (node:internal/modules/esm/resolve:845:9)
    at moduleResolve (node:internal/modules/esm/resolve:890:18)
    at Loader.defaultResolve [as _resolve] (node:internal/modules/esm/resolve:1004:11)
    at Loader.resolve (node:internal/modules/esm/loader:89:40)
    at Loader.getModuleJob (node:internal/modules/esm/loader:242:28)
    at Loader.import (node:internal/modules/esm/loader:177:28)
    at node:internal/process/esm_loader:57:31
    at initializeLoader (node:internal/process/esm_loader:62:5)
    at Object.loadESM (node:internal/process/esm_loader:67:11) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

Is there any other way I could do this?

**Comment by ljharb** [comment]


**Reply by Maxim-Mazurok** [reply]
It's the same command as I mentioned above, but with short arguments and single quotes.
And it doesn't work if I don't have `ts-node` installed in the cwd or globally.

**Reply by Maxim-Mazurok** [reply]
You can try `npm rm ts-node && npm rm -g ts-node && npx -p ts-node -c 'node --loader ts-node/esm'` to see that it'll fail

**Reply by ljharb** [reply]
ah, my mistake, i didn't realize `--call` was the long form of `-c`.

**Reply by Maxim-Mazurok** [reply]
I learned from [here](https://github.com/npm/npx/issues/5#issuecomment-689306306) that npx supports all npm arguments, including `--node-options`, so I've tried this:
`npx -y --package ts-node --node-options "--loader ts-node/esm --experimental-specifier-resolution=node" --call "node ./tools/not-ignored-files.ts"`
But it also didn't work, same error.

**Comment by MylesBorins** [comment, marked as answer]
@Maxim-Mazurok the Node.js loader algorithm cannot resolve global packages today. Even if you set your NODE_PATH to include the globally installed folder.

There is nothing that npm or npx can do to fix this at the moment. Your best bet is to install ts-node in the package you are looking to do this with

**Reply by Maxim-Mazurok** [reply]
That would defeat the purpose of using npx, but thank you for the answer. I'll hope that my issue will be solved by publishing ts-node-esm package (https://github.com/TypeStrong/ts-node/issues/1258)
